### PR TITLE
Implementation of contenteditable

### PR DESCRIPTION
**Pull request by pavpanchekha**
This PR adds text and code to implement `contenteditable`. The implementation is fairly basic: you can only add text to the last text node inside the editable element, and spaces aren't handled properly (they are stripped so they don't move the cursor). But I think it suffices to demonstrate editing lag.

They also highlight some kind of bug, because even with invalidation, editing is pretty laggy. Something to investigate, but I don't think it blocks merge.

It's still missing tests but should be reviewable.

### CODE REVIEW

File: book/invalidation.md [hunk 1]
```diff
@@ -12,6 +12,118 @@ time in layout, not graphics. Yet just like compositing enabled smooth
...
+            ]
+            if text_nodes:
+                cx = text_nodes[-1].x + text_nodes[-1].width
+                cy1 = text_nodes[-1].y
+                cy2 = text_nodes[-1].y + text_nodes[-1].height
+            else:
+                cx = self.x
+                cy1 = self.y
+                cy2 = self.y + self.height
+            cmds.append(DrawLine(cx, cy1, cx, cy2))
```
Comment (chrishtr) [inline book/invalidation.md:119]: Factor out similar code for `InputLayout` and call it from each place, to draw the reader's
attention to the shared concept?

File: book/invalidation.md [hunk 2]
```diff
@@ -12,6 +12,118 @@ time in layout, not graphics. Yet just like compositing enabled smooth
...
+        # ...
+```
+
+You should now be able to edit the example above in your own
+browser---but if you try it, you'll see that editing is extremely
+slow, with each character typed adding multiple frames of delay.
```
Comment (chrishtr) [inline book/invalidation.md:125]:
```suggestion
slow, with each character typed adding multiple 60Hz frames of delay.
```

File: book/invalidation.md [hunk 3]
```diff
@@ -899,12 +1023,11 @@ time a dirty flag is set, we'll want to set the `dirty_descendants`
 flag on all ancestors:
 
 ``` {.python replace=BlockLayout:/LineLayout:}
```
Comment (chrishtr) [inline book/invalidation.md:1025]:
```suggestion
``` {.python}
```

File: src/wbetools.py [hunk 4]
```diff
@@ -19,17 +20,24 @@ def decorator(new_cls):
                 assert obj is None, "Cannot patch using a closure"
                 assert getattr(existing_cls, attr) is None, "Cannot patch a closure"
             elif attr == "__globals__":
-                continue
                 old_obj = getattr(existing_cls, attr)
-                for field in obj:
-                    if field not in old_obj:
-                        print(f"While patching {existing_cls}, __globals__.{field} not in old")
+                new_name = obj["__name__"]
```
Comment (chrishtr) [inline src/wbetools.py:24]: What are the changes in this file doing?

Comment (chrishtr) [reply src/wbetools.py:24]: Please add a comment explaining.

File: book/invalidation.md [hunk 5]
```diff
@@ -12,6 +12,118 @@ time in layout, not graphics. Yet just like compositing enabled smooth
...
+
+One good example is editing text. Most people type at a rate of
+several characters per second, and even a delay of a few frames is
+very distracting. Yet editing text changes the document, and therefore
+requires constructing a new layout tree that reflects the new text.
+And, in fact, if you open up a large web page (like this one) and type
+into an input box on it (like the one below) in your browser, you'll
+see that it is quite laggy indeed:
+
+<input style="width:100%"/>
```
[the remark this answers is not in the document]
Comment (pavpanchekha) [reply book/invalidation.md:36]: I don't think it understands any other width measure either. It's OK if it just applies to real browsers.

File: infra/runtests.py [hunk 6]
```diff
@@ -40,10 +40,14 @@ def run_tests(chapter, file_name):
     # This ugly code reloads all of our modules from scratch, in case
     # a test makes a mutation to a global for some reason
     src_dir = os.path.split(os.path.realpath(file_name))[0]
-    for name, mod in list(sys.modules.items()):
+    mods = list(sys.modules.items())
+    for name, mod in sorted(mods, key=lambda x: (len(x[0]), x[0])):
         if hasattr(mod, "__file__") and mod.__file__ and \
            os.path.realpath(mod.__file__).startswith(src_dir) and \
            mod.__file__.endswith("py"):
+            for attr in dir(mod):
```
Comment (chrishtr) [inline infra/runtests.py:48]: What does this code do?

File: src/wbetools.py [hunk 7]
```diff
@@ -19,17 +20,32 @@ def decorator(new_cls):
...
+                # might use a global variable (normally, another
+                # function) that's defined in the new scope but not
+                # the old one. To prevent this, we copy the new
+                # function into the old scope, as long as it's not
+                # already defined. (If it's already defined, we throw
+                # an error.) We only do this for variables that are
+                # actually used; to get that, we walk the bytecode
+                # looking for LOAD_GLOBAL operations.
                 old_obj = getattr(existing_cls, attr)
-                for field in obj:
```
Comment (chrishtr) [inline src/wbetools.py:24]: Is this old code no longer needed?

File: infra/runtests.py (hunk 6 above)
Comment (pavpanchekha) [reply infra/runtests.py]: I refactored it to be clearer. Basically: it reloads all of *our* modules, in order.

**Review by chrishtr — APPROVED** [review]

**Pull request merged**